File: scenarios/pravega-flink-connector-sql-samples/src/main/java/io/pravega/connectors/nytaxi/MaxTravellersPerDestination.java
/*
 * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 */
package io.pravega.connectors.nytaxi;

import lombok.extern.slf4j.Slf4j;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.table.api.Table;
import org.apache.flink.table.api.bridge.java.StreamTableEnvironment;
import org.apache.flink.table.api.Tumble;
import org.apache.flink.types.Row;

import static org.apache.flink.table.api.Expressions.$;
import static org.apache.flink.table.api.Expressions.lit;

/**
 * Find maximum number of travellers who travelled to a destination point for a given window interval.
 */

@Slf4j
public class MaxTravellersPerDestination extends AbstractHandler {

    public MaxTravellersPerDestination (String ... args) {
        super(args);
    }

    @Override
    public void handleRequest() {

        StreamExecutionEnvironment env = getStreamExecutionEnvironment();

        // create a TableEnvironment
        StreamTableEnvironment tEnv = StreamTableEnvironment.create(env);

        tEnv.executeSql(createTableDdl("WATERMARK FOR dropOffTime AS dropOffTime - INTERVAL '30' SECONDS", "max-traveller"));

        Table noOfTravelersPerDest = tEnv
                .from("TaxiRide")
                .select($("passengerCount"), $("dropOffTime"), $("destLocationZone"))
                .window(Tumble.over(lit(1).hour()).on($("dropOffTime")).as("w"))
                .groupBy($("destLocationZone"), $("w"))
                .select($("destLocationZone"), $("w").start().as("start"), $("w").end().as("end"), $("passengerCount").count().as("cnt"));

        tEnv.toAppendStream(noOfTravelersPerDest, Row.class).print();

        try {
            env.execute("Max-Travellers-Per-Destination");
        } catch (Exception e) {
            log.error("Application Failed", e);
        }
    }
}

File: scenarios/pravega-flink-connector-sql-samples/src/main/python/max_travellers_per_destination.py
from pyflink.common.typeinfo import Types
from pyflink.datastream import StreamExecutionEnvironment
from pyflink.table.expressions import col
from pyflink.table.table_environment import StreamTableEnvironment
from pyflink.table.window import Tumble

from common import create_table_ddl


def max_travellers_per_destination():
    env = StreamExecutionEnvironment.get_execution_environment()
    t_env = StreamTableEnvironment.create(stream_execution_environment=env)

    t_env.execute_sql(create_table_ddl(
            "WATERMARK FOR dropOffTime AS dropOffTime - INTERVAL '30' SECONDS"))
    taxi_ride = t_env.from_path('TaxiRide')
    no_of_travelers_per_dest = taxi_ride \
        .select(taxi_ride.passengerCount, taxi_ride.dropOffTime, taxi_ride.destLocationZone) \
        .window(Tumble().over('1.hour').on(taxi_ride.dropOffTime).alias('w')) \
        .group_by(taxi_ride.destLocationZone, col('w')) \
        .select(taxi_ride.destLocationZone, \
                col('w').start.alias('start'), \
                col('w').end.alias('end'), \
                taxi_ride.passengerCount.count.alias('cnt'))

    t_env.to_append_stream(
        no_of_travelers_per_dest,
        Types.ROW_NAMED(['destLocationZone', 'start', 'end', 'cnt'],
            [Types.INT(),
             Types.SQL_TIMESTAMP(),
             Types.SQL_TIMESTAMP(),
             Types.LONG()])).print()


if __name__ == '__main__':
    max_travellers_per_destination()
